### ISSUE TITLE
Column types are wrong with larger scan depth and empty values, how to handle empty values

### ISSUE THREAD

**Issue by papyr** [Cinchoo/ChoETL#73]
Hello I am using choETL version 1.0.9.9. I like your lib. a lot 🥇  since it offers _Data type discovery, some of this does not work with new releases so I opted to stay at that version_. 

I need help in configuring this correctly _to handle Type discovery even if there are empty values_.

**The `Cost1` `Cost2`,.`.3.` columns data type, shows up wrongly as `String`, when it should be `double`/number.** using `WithMaxScanRows(6)`,  but if I change it to `WithMaxScanRows(3)` it works fine.

    using (var choReader = new ChoCSVReader(someCsvFile).WithFirstLineHeader().WithMaxScanRows(3))
             {   
                 var   dataTable = choReader.AsDataTable();

                 foreach (DataColumn column in table.Columns)              
                  var checkType =  SQLGetType(column) ; // if scan rows is 3, it works fine, but if its 6 i get string, 
                  // how can I ensure it gives me the proper dataype even if there are empty data cells.
                 
                ...
             }

But when I set `WithMaxScanRows(3)` the `Cost1..2..3.` show up correctly as double.
________________________
CSV file

Date,Project,Login,Area,**Cost1**,`Cost2`,`Cost3`,Cost4,Cost5
2018-03-19 23:00:00,jasonLogin,sr1,taskscompleted,21.0,0.0,0.0,0.0,0.0
2018-06-27 23:00:00,jasonLogin,bblackmon,aircraftdelivery,480.0,0.0,0.0,0.0,0.0
2017-12-19 23:00:00,jasonLogin,sfinishing1,hoursearned,94.73,0.23333333333333334,0.0,0.0,0.0
2018-03-27 23:00:00,jasonLogin,bblackmon,hoursearned,46.5,0.0,0.0,0.0,0.0
2017-12-05 23:00:00,jasonLogin,tmodrick,toolissues,,,,,0.0
2018-03-11 23:00:00,jasonLogin,jclark,hoursearned,72.39,0.0,0.0,0.0,0.0
2018-02-28 23:00:00,jasonLogin,jmartinez,housekeeping1,30.0,0.0,0.0,0.0,0.0
2018-02-17 23:00:00,jasonLogin,sfinishing1,budgetrevenue,2018.94,0.0,0.0,0.0,0.0
2018-05-30 23:00:00,jasonLogin,khall,hoursworked,45.7,0.0,0.0,0.0,0.0
2018-03-28 23:00:00,jasonLogin,finish1,budgetrevenue,4834.77,0.0,0.0,0.0,0.0
2017-12-15 23:00:00,jasonLogin,btwilley,aircraftmove1,,,,,0.0

**Comment by Cinchoo** [comment]
Found the issue, putting a fix. Will be releasing new package (1.1.0.5). Try and let me know.

**Comment by papyr** [comment]
ok thanks, will test it and let you know.

 How do I check if file is empty, i.e. where _first row is header_, but no data after that.

**Comment by Cinchoo** [comment]
Please take nuget package version (1.1.0.5-beta5), try the fix.

1. How do I check if file is empty, i.e. where first row is header, but no data after that.

```
using (var p = ChoCSVReader.LoadText(csv)
    .WithFirstLineHeader()
    )
{
    var count = p.Count();
}
```

2. How to override fields during discovery

```
using (var p = ChoCSVReader.LoadText(csv)
    .WithFirstLineHeader()
    .WithMaxScanRows(2)
    )
{
    p.RecordFieldTypeAssessment += (o, e) =>
    {
        e.FieldTypes["Id"] = typeof(short);
    };
    foreach (var rec in p)
        Console.WriteLine(rec.Dump());
}

```

**Comment by papyr** [comment]
Thanks.

I dont see the beta version in nuget..  does Not show '1.1.0.5-beta5' only 1.1.0.4

![image](https://user-images.githubusercontent.com/12733414/73868749-098e7480-4817-11ea-9430-1dd58fe137c2.png)

**Comment by Cinchoo** [comment]
On the top 'Include Prerelease' checkbox, check that box.

![image](https://user-images.githubusercontent.com/9448455/73869527-68082280-4818-11ea-881f-a2ea327386ab.png)

**Comment by papyr** [comment]
Works now!

**Comment by papyr** [comment]
Hello @Cinchoo , the issue is happening intermittently again with larger datasets if I run the same data file 100 times it fails 5 times. So, let me scrub and send some data to you.

Also I noticed the releases are no longer tagged here. I would like to help with debugging, how can I go about pulling the code for that _specific_ beta version specifically and test. 

Lastly, am I allowed to modify the code so it fits our use case.

Thanks

**Comment by papyr** [comment]
If there are 50 cols, this is getting get & set everytime. 
```
p.RecordFieldTypeAssessment += (o, e) =>
    {
        e.FieldTypes["Id"] = typeof(short);
    };
```
____________________
Rather is it possible to get just the `list of DataTypes`, so we can enumerate over that. I was going to try/make these changes not sure if its already there as a different function. 

`p.GetHeaderTypes.ToList();`

**Comment by Cinchoo** [comment]
well, `RecordFieldTypeAssessment` is called for each record, to take control of assessing the field type from its contents.

On the other hand, there is another event `MembersDiscovered`, this get called once where you can assign field types without considering the payload. 

```
p.MembersDiscovered += (o, e) =>
{
    var ft = e.Value; 

    ft["Id.x"] = typeof(short);
    ft["Name"] = typeof(string);
    ft["City"] = typeof(string);
};
```

PS. found a bug that this method get called twice in the life of CSV load. Put a fix in 1.1.0.5-beta6 package.
